### PR TITLE
US5213 policy relations

### PR DESCRIPTION
**Pull request by krimsonkla**

### CODE REVIEW

File: app/controllers/mongoidable/abilities_controller.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+# TODO: Ensure Mongoid railtie rescue_responses handles DocumentNotFound, Validations
+# TODO: For cancan add the dispatch rescue
+# config.action_dispatch.rescue_responses.merge!(
+#   'ActiveRecord::RecordNotFound'   => :not_found,
+#   'ActiveRecord::StaleObjectError' => :conflict,
+#   'ActiveRecord::RecordInvalid'    => :unprocessable_entity,
+#   'ActiveRecord::RecordNotSaved'   => :unprocessable_entity
+# )
+module Mongoidable
```
Comment (krimsonkla) [inline app/controllers/mongoidable/abilities_controller.rb:11]: Copied and and massaged from MM

File: app/controllers/mongoidable/policies_controller.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# Provides a CRUD api for ability policies.
+module Mongoidable
```
Comment (krimsonkla) [inline app/controllers/mongoidable/policies_controller.rb:4]: Copied and and massaged from MM

File: app/models/mongoidable/ability.rb [hunk 3]
```diff
@@ -43,6 +43,20 @@ def ==(other)
         other.extra == extra
     end
 
+    def merge_requirements(data)
+      return if @merged
```
Comment (krimsonkla) [inline app/models/mongoidable/ability.rb:47]: Merges the required extra cancan arguments into the policies instance abilities

File: app/models/mongoidable/policy.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module Mongoidable
```
Comment (krimsonkla) [inline app/models/mongoidable/policy.rb:3]: Copied and and massaged from MM

File: app/models/mongoidable/policy_relation.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+module Mongoidable
```
Comment (krimsonkla) [inline app/models/mongoidable/policy_relation.rb:3]: An embedded model that works as a has policy through type relation. It contains the required attributes in order to apply the policy to the given model.

File: app/queries/mongoidable/ability_query.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module Mongoidable
```
Comment (krimsonkla) [inline app/queries/mongoidable/ability_query.rb:3]: Copied and and massaged from MM

File: app/queries/mongoidable/concerns/updates_abilities.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
```
Comment (krimsonkla) [inline app/queries/mongoidable/concerns/updates_abilities.rb:2]: Copied and and massaged from MM

File: app/queries/mongoidable/policy_query.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module Mongoidable
```
Comment (krimsonkla) [inline app/queries/mongoidable/policy_query.rb:3]: Copied and and massaged from MM

File: app/serializers/mongoidable/concerns/serializes_ability_data.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
```
Comment (krimsonkla) [inline app/serializers/mongoidable/concerns/serializes_ability_data.rb:2]: Copied and and massaged from MM

File: app/serializers/mongoidable/concerns/serializes_casl_abilities.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
```
Comment (krimsonkla) [inline app/serializers/mongoidable/concerns/serializes_casl_abilities.rb:1]: Copied and and massaged from MM

File: app/serializers/mongoidable/concerns/serializes_instance_abilities.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
```
Comment (krimsonkla) [inline app/serializers/mongoidable/concerns/serializes_instance_abilities.rb:2]: Copied and and massaged from MM

File: app/serializers/mongoidable/policy_serializer.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
```
Comment (krimsonkla) [inline app/serializers/mongoidable/policy_serializer.rb:2]: Copied and and massaged from MM

File: lib/mongoidable/class_abilities.rb [hunk 13]
```diff
@@ -26,6 +26,13 @@ def inherits_from
         @inherits_from ||= superclass.respond_to?(:inherits_from) ? superclass.inherits_from.dup : []
       end
 
+      def accepts_policies_as(name)
```
Comment (krimsonkla) [inline lib/mongoidable/class_abilities.rb:29]: class method for defining policy relationships. Policy abilities are transposed through the relationships onto the parent model. 

Example
```
class User
  accepts_policies_as :organization_policies
end

u = User.new
u.organization_policies.create(requirements: {}, policy: {})
```

**Review by RealNobody — COMMENTED** [review]
A few small things, but the `hash_with_indifferent_access` will be a problem.

File: app/serializers/mongoidable/policy_serializer.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+module Mongoidable
+  # Default serializer for policies
+  class PolicySerializer < ActiveModel::Serializer
+    include Concerns::SerializesInstanceAbilities
+
+    attributes :_id, :name, :description, :owner_type, :requirements
```
[the remark this answers is not in the document]
Comment (krimsonkla) [reply app/serializers/mongoidable/policy_serializer.rb:8]: No tiny ID in what will be concidered a 3rd party gem. That will be in the derived policy class in MM.

File: app/queries/mongoidable/ability_query.rb [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+          remove: unsafe_params[:remove_policy]).call
+      else
+        abilities = Array.wrap(unsafe_params[:instance_ability] || unsafe_params[:instance_abilities])
+        Mongoidable::AbilitiesUpdater.new(object_for_update, abilities, replace: unsafe_params[:replace]).call
+      end
+    end
+
+    def unsafe_params
+      params.to_unsafe_hash
+    end
```
Comment (jaurand) [inline app/queries/mongoidable/ability_query.rb:50]: This sounds... unsafe

**Review by jaurand — APPROVED** [review]

**Review by RealNobody — APPROVED** [review]

**Pull request merged**